gobblin-modules/gobblin-gitter/src/main/java/org/apache/gobblin/source/GitterMessageSource.java

```diff
@@ -0,0 +1,132 @@
+ */
+@Slf4j
+public class GitterMessageSource extends AbstractSource<Schema, GenericRecord> {
+
+  @Override
+  public List<WorkUnit> getWorkunits(SourceState state) {
+    Map<String, Iterable<WorkUnitState>> previousWorkUnits = state.getPreviousWorkUnitStatesByDatasetUrns();
+    List<String> gitterRooms = new LinkedList<>(Splitter.on(",").omitEmptyStrings().
+        splitToList(state.getProp(GitterMessageExtractor.GITTER_ROOM_NAMES_KEY)));
+
+    Map<String, BigIntegerWatermark> prevHighWatermarks = Maps.newHashMap();
+    for (Map.Entry<String, Iterable<WorkUnitState>> entry : previousWorkUnits.entrySet()) {
+      Iterable<BigIntegerWatermark> watermarks =
+          Iterables.transform(entry.getValue(), new Function<WorkUnitState, BigIntegerWatermark>() {
+            @Override
+            public BigIntegerWatermark apply(WorkUnitState wus) {
+              return wus.getActualHighWatermark(BigIntegerWatermark.class);
+            }
+          });
+      watermarks = Iterables.filter(watermarks, Predicates.notNull());
+      List<BigIntegerWatermark> watermarkList = Lists.newArrayList(watermarks);
+      if (watermarkList.size() > 0) {
+        prevHighWatermarks.put(entry.getKey(), Collections.max(watermarkList));
+      }
+    }
+
+    Extract extract = createExtract(Extract.TableType.SNAPSHOT_ONLY,
+        state.getProp(ConfigurationKeys.EXTRACT_NAMESPACE_NAME_KEY), "GitterOutput");
+    List<WorkUnit> workUnits = Lists.newArrayList();
+
+    for (String gitterRoom : gitterRooms) {
+      BigIntegerWatermark prevWatermark = prevHighWatermarks.containsKey(gitterRoom) ? prevHighWatermarks.get(gitterRoom) :
+          new BigIntegerWatermark(BigInteger.ZERO);
+      prevHighWatermarks.remove(gitterRoom);
+      log.info(String.format("Creating workunit for room: %s with low watermark: %s", gitterRoom,
+          prevWatermark.getValue().toString(16)));
+      WorkUnit workUnit = WorkUnit.create(extract, new WatermarkInterval(prevWatermark,
+          new BigIntegerWatermark(BigInteger.ZERO)));
+      workUnit.setProp(ConfigurationKeys.DATASET_URN_KEY, gitterRoom);
+      workUnits.add(workUnit);
+    }
+
+    for (Map.Entry<String, BigIntegerWatermark> nonProcessedDataset : prevHighWatermarks.entrySet()) {
+      WorkUnit workUnit = WorkUnit.create(extract, new WatermarkInterval(nonProcessedDataset.getValue(),
+          nonProcessedDataset.getValue()));
+      workUnit.setProp(ConfigurationKeys.DATASET_URN_KEY, nonProcessedDataset.getKey());
+      workUnits.add(workUnit);
+    }
+
+    return workUnits;
+  }
+
+  @Override
+  public Extractor<Schema, GenericRecord> getExtractor(WorkUnitState state)
+      throws IOException {
+    ClassAliasResolver<Extractor> aliasResolver = new ClassAliasResolver<>(Extractor.class);
+    Extractor<Schema, GenericRecord> extractor = null;
+    try {
+      String extractorClass = state.getProp(GITTER_EXTRACTOR_KEY, GITTER_EXTRACTOR_DEFAULT);
```

Comment:
How do we guarantee the extractor is compatible with the source? Shall we make sure it was extended from some base gitter extractor?